### PR TITLE
Added Typescript example for withRouter

### PR DESCRIPTION
**Pull request by faizanv**
I added an example of extending props to use `withRouter` in TypeScript. I couldn't find anything online besides this [Github Issue](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/24077) so I just emulated that. Let me know if an actual interface already exists that can be imported and extended instead.

### CODE REVIEW

File: docs/api-reference/next/router.md [hunk 1]
```diff
@@ -392,3 +392,26 @@ function Page({ router }) {
...
 export default withRouter(Page)
 ```
+
+### Typescript
+
+To use class components with `withRouter`, the component needs to accept a router prop like so
+
+```tsx
+import React from 'react'
+import { withRouter, SingletonRouter } from 'next/router'
```
Comment (timneutkens) [inline docs/api-reference/next/router.md:402]:
```suggestion
import { withRouter, NextRouter } from 'next/router'
```

File: docs/api-reference/next/router.md [hunk 2]
```diff
@@ -392,3 +392,26 @@ function Page({ router }) {
...
+
+interface WithRouterComponent {
+  router: SingletonRouter
```
Comment (timneutkens) [inline docs/api-reference/next/router.md:405]:
```suggestion
  router: NextRouter
```

File: docs/api-reference/next/router.md [hunk 3]
```diff
@@ -392,3 +392,26 @@ function Page({ router }) {
...
+
+### Typescript
+
+To use class components with `withRouter`, the component needs to accept a router prop like so
+
+```tsx
+import React from 'react'
+import { withRouter, SingletonRouter } from 'next/router'
+
+interface WithRouterComponent {
```
Comment (timneutkens) [inline docs/api-reference/next/router.md:404]:
```suggestion
interface WithRouterProps {
```

File: docs/api-reference/next/router.md [hunk 4]
```diff
@@ -392,3 +392,26 @@ function Page({ router }) {
...
+  router: SingletonRouter
+}
+
+interface MyComponentProps extends WithRouterComponent {}
```
Comment (timneutkens) [inline docs/api-reference/next/router.md:408]:
```suggestion
interface MyComponentProps extends WithRouterProps {}
```

File: docs/api-reference/next/router.md [hunk 5]
```diff
@@ -392,3 +392,26 @@ function Page({ router }) {
 
 export default withRouter(Page)
 ```
+
+### Typescript
+
+To use class components with `withRouter`, the component needs to accept a router prop like so
```
Comment (timneutkens) [inline docs/api-reference/next/router.md:398]:
```suggestion
To use class components with `withRouter`, the component needs to accept a router prop:
```

**Review by timneutkens — APPROVED** [review]

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 7aa33872ee01e9cac864d08c077c941d2752c724

**test/integration/dist-dir/test/index.test.js** 
- distDir > should handle null/undefined distDir

<details>
<summary>Expand output</summary>

  ● distDir › should handle null/undefined distDir

    expect(received).toBe(expected) // Object.is equality

    Expected: 0
    Received: 88

      63 |     await fs.writeFile(nextConfig, origNextConfig)
      64 | 
    > 65 |     expect(stderr.length).toBe(0)
         |                           ^
      66 |   })
      67 | })
      68 | 

      at Object.<anonymous> (integration/dist-dir/test/index.test.js:65:27)

</details>

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| buildDuration | 11.1s | 10.8s | -287ms |
| nodeModulesSize | 73.3 MB | 73.3 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.179 | 2.071 | -0.11 |
| / avg req/sec | 1147.43 | 1207.38 | +59.95 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.423 | 1.35 | -0.07 |
| /error-in-render avg req/sec | 1757.42 | 1851.49 | +94.07 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| 677f882d2ed8..38ed.js gzip | 13.1 kB | 13.1 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-HASH.js gzip | 6.63 kB | 6.63 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59.4 kB | 59.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.3 kB | 31.3 kB | ✓ |
| Overall change | 31.3 kB | 31.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| _app-2a09aa2..4a98.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-8b758..aef6.js gzip | 3.46 kB | 3.46 kB | ✓ |
| hooks-c71ae4..70cd.js gzip | 887 B | 887 B | ✓ |
| index-bbee2f..528b.js gzip | 227 B | 227 B | ✓ |
| link-7faf09b..eba4.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..bf84.js gzip | 303 B | 303 B | ✓ |
| withRouter-a..5826.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.09 kB | 8.09 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| _buildManifest.js gzip | 321 B | 321 B | ✓ |
| Overall change | 321 B | 321 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| index.html gzip | 614 B | 614 B | ✓ |
| link.html gzip | 620 B | 620 B | ✓ |
| withRouter.html gzip | 608 B | 608 B | ✓ |
| Overall change | 1.84 kB | 1.84 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| buildDuration | 12.7s | 12.5s | -167ms |
| nodeModulesSize | 73.3 MB | 73.3 MB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| 677f882d2ed8..38ed.js gzip | 13.1 kB | 13.1 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-HASH.js gzip | 6.63 kB | 6.63 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59.4 kB | 59.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.3 kB | 31.3 kB | ✓ |
| Overall change | 31.3 kB | 31.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| _app-2a09aa2..4a98.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-8b758..aef6.js gzip | 3.46 kB | 3.46 kB | ✓ |
| hooks-c71ae4..70cd.js gzip | 887 B | 887 B | ✓ |
| index-bbee2f..528b.js gzip | 227 B | 227 B | ✓ |
| link-7faf09b..eba4.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..bf84.js gzip | 303 B | 303 B | ✓ |
| withRouter-a..5826.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.09 kB | 8.09 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| _buildManifest.js gzip | 321 B | 321 B | ✓ |
| Overall change | 321 B | 321 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| _error.js | 1 MB | 1 MB | ✓ |
| 404.html | 2.67 kB | 2.67 kB | ✓ |
| hooks.html | 1.92 kB | 1.92 kB | ✓ |
| index.js | 1 MB | 1 MB | ✓ |
| link.js | 1.06 MB | 1.06 MB | ✓ |
| routerDirect.js | 1.05 MB | 1.05 MB | ✓ |
| withRouter.js | 1.05 MB | 1.05 MB | ✓ |
| Overall change | 5.19 MB | 5.19 MB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Webpack 5 Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| buildDuration | 12.9s | 12.8s | -149ms |
| nodeModulesSize | 73.3 MB | 73.3 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.128 | 2.05 | -0.08 |
| / avg req/sec | 1174.98 | 1219.77 | +44.79 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.267 | 1.332 | ⚠️ +0.07 |
| /error-in-render avg req/sec | 1973.07 | 1876.72 | ⚠️ -96.35 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| 597-c48889ef..1870.js gzip | 13 kB | 13 kB | ✓ |
| framework.HASH.js gzip | 39.3 kB | 39.3 kB | ✓ |
| main-HASH.js gzip | 6.58 kB | 6.58 kB | ✓ |
| webpack-HASH.js gzip | 954 B | 954 B | ✓ |
| Overall change | 59.8 kB | 59.8 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.1 kB | 31.1 kB | ✓ |
| Overall change | 31.1 kB | 31.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| _app-a5b2e84..33cf.js gzip | 1.26 kB | 1.26 kB | ✓ |
| _error-6f1f9..5577.js gzip | 3.38 kB | 3.38 kB | ✓ |
| hooks-725de8..0a15.js gzip | 904 B | 904 B | ✓ |
| index-939503..6e1c.js gzip | 232 B | 232 B | ✓ |
| link-e0cc871..fdbb.js gzip | 1.63 kB | 1.63 kB | ✓ |
| routerDirect..9360.js gzip | 308 B | 308 B | ✓ |
| withRouter-6..44ec.js gzip | 304 B | 304 B | ✓ |
| Overall change | 8.02 kB | 8.02 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| _buildManifest.js gzip | 299 B | 299 B | ✓ |
| Overall change | 299 B | 299 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | faizanv/next.js faizanv-router-types | Change |
| - | - | - | - |
| index.html gzip | 588 B | 588 B | ✓ |
| link.html gzip | 593 B | 593 B | ✓ |
| withRouter.html gzip | 581 B | 581 B | ✓ |
| Overall change | 1.76 kB | 1.76 kB | ✓ |

</details>

#### Diffs
<details>
<summary>Diff for <strong>index.html</strong></summary>

```diff
@@ -43,7 +43,7 @@
         "props": { "pageProps": {} },
         "page": "/",
         "query": {},
-        "buildId": "5rQrMWYen8LzYXNdYgW6T",
+        "buildId": "PFeZ6nzHShEtlu_li41Lu",
         "isFallback": false,
         "gip": true
       }
@@ -77,11 +77,11 @@
       async=""
     ></script>
     <script
-      src="/_next/static/5rQrMWYen8LzYXNdYgW6T/_buildManifest.js"
+      src="/_next/static/PFeZ6nzHShEtlu_li41Lu/_buildManifest.js"
       async=""
     ></script>
     <script
-      src="/_next/static/5rQrMWYen8LzYXNdYgW6T/_ssgManifest.js"
+      src="/_next/static/PFeZ6nzHShEtlu_li41Lu/_ssgManifest.js"
       async=""
     ></script>
   </body>
```
</details>
<details>
<summary>Diff for <strong>link.html</strong></summary>

```diff
@@ -48,7 +48,7 @@
         "props": { "pageProps": {} },
         "page": "/link",
         "query": {},
-        "buildId": "5rQrMWYen8LzYXNdYgW6T",
+        "buildId": "PFeZ6nzHShEtlu_li41Lu",
         "isFallback": false,
         "gip": true
       }
@@ -82,11 +82,11 @@
       async=""
     ></script>
     <script
-      src="/_next/static/5rQrMWYen8LzYXNdYgW6T/_buildManifest.js"
+      src="/_next/static/PFeZ6nzHShEtlu_li41Lu/_buildManifest.js"
       async=""
     ></script>
     <script
-      src="/_next/static/5rQrMWYen8LzYXNdYgW6T/_ssgManifest.js"
+      src="/_next/static/PFeZ6nzHShEtlu_li41Lu/_ssgManifest.js"
       async=""
     ></script>
   </body>
```
</details>
<details>
<summary>Diff for <strong>withRouter.html</strong></summary>

```diff
@@ -43,7 +43,7 @@
         "props": { "pageProps": {} },
         "page": "/withRouter",
         "query": {},
-        "buildId": "5rQrMWYen8LzYXNdYgW6T",
+        "buildId": "PFeZ6nzHShEtlu_li41Lu",
         "isFallback": false,
         "gip": true
       }
@@ -77,11 +77,11 @@
       async=""
     ></script>
     <script
-      src="/_next/static/5rQrMWYen8LzYXNdYgW6T/_buildManifest.js"
+      src="/_next/static/PFeZ6nzHShEtlu_li41Lu/_buildManifest.js"
       async=""
     ></script>
     <script
-      src="/_next/static/5rQrMWYen8LzYXNdYgW6T/_ssgManifest.js"
+      src="/_next/static/PFeZ6nzHShEtlu_li41Lu/_ssgManifest.js"
       async=""
     ></script>
   </body>
```
</details>
</details>
Commit: 7aa33872ee01e9cac864d08c077c941d2752c724

**Pull request merged**